File: leetcode/155.Min-Stack.py
class Item:
    def __init__(self, val: int, last_min: int):
        self.val = val
        self.last_min = last_min

"""
https://leetcode.com/problems/min-stack/description/

Time complexity: all O(1)
Space complexity: O(n)
"""
class MinStack:

    stk: List[Item]

    def __init__(self):
        self.stk = []

    def push(self, val: int) -> None:
        self.stk.append(Item(val, val if len(self.stk) == 0 else min(self.stk[-1].last_min, val)))

    def pop(self) -> None:
        self.stk.pop()

    def top(self) -> int:
        return self.stk[-1].val

    def getMin(self) -> int:
        return self.stk[-1].last_min

File: engineering/LeetCode/155.min-stack.ts
/*
 * @lc app=leetcode id=155 lang=typescript
 *
 * [155] Min Stack
 */

// @lc code=start

class MinStack {

    arr: Array<[number, number]>;

    constructor() { this.arr = []; }

    push(val: number): void {
        if (this.arr.length === 0) { this.arr.push([val, val]); return; }
        const lastMin = this.getMin();
        this.arr.push(val < lastMin ? [val, val] : [val, lastMin]);
    }

    pop(): void {
        this.arr.pop();
    }

    top(): number { return this.arr[this.arr.length - 1][0]; }

    getMin(): number { return this.arr[this.arr.length - 1][1]; }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
// @lc code=end